### PR TITLE
Remove QMCMainState and CoupledMC

### PR DESCRIPTION
**Pull request by williamfgc**
## Proposed changes

Simplify the hierarchy and clean QMCMain
Remove QMCMainState and CoupledMC file integrating everything into QMCMain
Move pools scope from public to private in QMCMain
Related to #3411 #4099

## What type(s) of changes does this code introduce?

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
sulfur with minimal configuration, must pass CI

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by williamfgc** [comment]
Test this please

File: src/QMCApp/qmcapp.cpp [hunk 1]
```diff
@@ -220,7 +220,7 @@ int main(int argc, char** argv)
     timingDoc.newDoc("resources");
     output_hardware_info(qmcComm, timingDoc, timingDoc.getRoot());
     timer_manager.output_timing(qmcComm, timingDoc, timingDoc.getRoot());
-    qmc->ptclPool->output_particleset_info(timingDoc, timingDoc.getRoot());
+    qmc->getParticlePool().output_particleset_info(timingDoc, timingDoc.getRoot());
```
Comment (ye-luo) [inline src/QMCApp/qmcapp.cpp:223]: This line reminds me the blurry boundary betwen qmcapp and QMCMain.
We will need to follow the following direction as we clean up old code.
qmcpapp is a layer that handles ensemble job.
All the timer manager, info.xml writing should be moved inside QMCMain.

File: src/QMCApp/QMCMain.h [hunk 2]
```diff
@@ -33,19 +33,34 @@ namespace qmcplusplus
...
+
+  /// QMCHamiltonian Pool
+  std::unique_ptr<HamiltonianPool> ham_pool_;
+
+  /// current MCWalkerConfiguration
+  MCWalkerConfiguration* qmc_system_;
+
+  ///Global estimators defined outside of <qmc> nodes
+  std::optional<EstimatorManagerInput> estimator_manager_input_;
+
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.h:63]: The MainState was introduced to improve the organization of member objects of QMCMain.
The down side is it hurt readability and having QMCMain as a base is awkward.
Can we have an anonymous struct here?
```
// QMCMainState
struct {
    /// ParticleSet Pool
  std::unique_ptr<ParticleSetPool> particle_set_pool_;
 ...
};
```
Maybe we can have a bit more organization this way.

Comment (ye-luo) [reply src/QMCApp/QMCMain.h:63]: Cannot use anonymous struct.
```
   51 |   std::unique_ptr<ParticleSetPool> particle_set_pool_;
      |                                    ^~~~~~~~~~~~~~~~~~
/home/yeluo/opt/qmcpack/src/QMCApp/QMCMain.h:51:36: error: member ‘std::unique_ptr<qmcplusplus::ParticleSetPool> qmcplusplus::QMCMain::<unnamed struct>::particle_set_pool_’ with destructor not allowed in anonymous aggregate
/home/yeluo/opt/qmcpack/src/QMCApp/QMCMain.h:51:36: error: member ‘std::unique_ptr<qmcplusplus::ParticleSetPool> qmcplusplus::QMCMain::<unnamed struct>::particle_set_pool_’ with copy assignment operator not allowed in anonymous aggregate
/home/yeluo/opt/qmcpack/src/QMCApp/QMCMain.h:54:37: error: member ‘std::unique_ptr<qmcplusplus::WaveFunctionPool> qmcplusplus::QMCMain::<unnamed struct>::psi_pool_’ with constructor not allowed in anonymous aggregate
```
Then ignore my suggestion.

**Review by ye-luo — APPROVED** [review]

File: src/QMCApp/qmcapp.cpp (hunk 1 above)
Comment (williamfgc) [reply src/QMCApp/qmcapp.cpp:223]: Thanks, that helps!

**Pull request merged**